File: pubnub-java/src/main/java/com/pubnub/api/endpoints/objects_api/memberships/GetMemberships.java
package com.pubnub.api.endpoints.objects_api.memberships;

import com.pubnub.internal.endpoints.DelegatingEndpoint;
import com.pubnub.api.endpoints.objects_api.utils.Include;
import com.pubnub.api.endpoints.objects_api.utils.PNSortKey;
import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction;
import com.pubnub.api.endpoints.remoteaction.MappingRemoteAction;
import com.pubnub.api.models.consumer.objects.PNPage;
import com.pubnub.api.models.consumer.objects_api.membership.PNGetMembershipsResult;
import com.pubnub.internal.InternalPubNubClient;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.Collection;
import java.util.Collections;

@Setter
@Accessors(chain = true, fluent = true)
public class GetMemberships extends DelegatingEndpoint<PNGetMembershipsResult> {

    private String uuid;
    private Integer limit;
    private PNPage page;
    private String filter;
    private Collection<PNSortKey> sort = Collections.emptyList();
    private boolean includeTotalCount;
    private boolean includeCustom;
    private Include.PNChannelDetailsLevel includeChannel;

    public GetMemberships(InternalPubNubClient pubnub) {
        super(pubnub);
    }

    @Override
    protected ExtendedRemoteAction<PNGetMembershipsResult> createAction() {
        return new MappingRemoteAction<>(
                pubnub.getMemberships(
                        uuid,
                        limit,
                        page,
                        filter,
                        SetMemberships.toInternal(sort),
                        includeTotalCount,
                        includeCustom,
                        SetMemberships.toInternal(includeChannel)
                ),
                PNGetMembershipsResult::from);
    }
}

File: pubnub-kotlin/src/main/kotlin/com/pubnub/api/endpoints/objects/membership/GetMemberships.kt
package com.pubnub.api.endpoints.objects.membership

import com.pubnub.api.Endpoint
import com.pubnub.api.PubNub
import com.pubnub.api.endpoints.remoteaction.map
import com.pubnub.api.models.consumer.objects.membership.PNChannelMembershipArrayResult
import com.pubnub.internal.DelegatingEndpoint
import com.pubnub.internal.endpoints.objects.membership.GetMemberships
import com.pubnub.internal.endpoints.objects.membership.IGetMemberships

/**
 * @see [PubNub.getMemberships]
 */
class GetMemberships internal constructor(private val getMemberships: GetMemberships) :
    DelegatingEndpoint<PNChannelMembershipArrayResult>(), IGetMemberships by getMemberships {

    override fun createAction(): Endpoint<PNChannelMembershipArrayResult> = getMemberships.map(PNChannelMembershipArrayResult::from)
}
